## NA imputation 

In [1]:
import jusipy
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
import missingno as msno
import copy
from fast_imputation import fast_imputation

#### Loading...

In [2]:
all_points = pd.read_pickle('data/all_points.pkl')

In [3]:
all_points.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22107 entries, 0 to 19999
Data columns (total 6 columns):
lat         22107 non-null object
long        22107 non-null float64
country     22107 non-null object
year        22107 non-null object
positive    22107 non-null int64
random      22107 non-null int64
dtypes: float64(1), int64(2), object(3)
memory usage: 1.2+ MB


In [4]:
A_country = jusipy.country_features.All()

Loading WorldBank(/home/marios/Desktop/Personal/repos/JuSiPy_Asser/jusipy/country_features/data/SD_API_15_DS2_en_csv_v2_10186395.csv)                                                                                                    

In [5]:
country_features = jusipy.country_features.get(all_points, A_country, fuzzy=50)

In [6]:
all_points2 = all_points.drop_duplicates(subset=['country','year'])

In [7]:
all_points2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 776 entries, 0 to 19844
Data columns (total 6 columns):
lat         776 non-null object
long        776 non-null float64
country     776 non-null object
year        776 non-null object
positive    776 non-null int64
random      776 non-null int64
dtypes: float64(1), int64(2), object(3)
memory usage: 42.4+ KB


In [8]:
country_features = jusipy.country_features.get(all_points2, A_country, fuzzy=50)

In [9]:
country_features.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 776 entries, 0 to 19844
Columns: 396 entries, WB_LGAF__WB_LGAF2016_1_1_1 to WB_SD__SP_DYN_LE00_MA_IN
dtypes: float64(280), object(116)
memory usage: 2.4+ MB


-------------------------------------------------------------------------------

### NA percentage (in red)

In [10]:
# somehow unconsciouly it came out like the French flag..

In [27]:
fields = ['WB_ARD__AG_SRF_TOTL_K2', 'WB_SE__WB_NY_GDP_MKTP_KD_ZG', 'WB_ARD__SP_RUR_TOTL', 'WB_ARD__EG_ELC_ACCS_RU_ZS', 'WB_ARD__AG_LND_TOTL_RU_K2',
          'WB_ARD__AG_LND_FRST_K2', 'WB_ARD__AG_LND_ARBL_ZS', 'WB_ARD__AG_LND_AGRI_K2', 'WB_ARD__AG_CON_FERT_ZS','WB_ARD__AG_LND_PRCP_MM','WB_ARD__EN_ATM_METH_AG_KT_CE',
         'WB_ARD__EN_ATM_NOXE_AG_KT_CE','WB_ARD__NV_AGR_TOTL_ZS','WB_ARD__TM_VAL_AGRI_ZS_UN','WB_ARD__TX_VAL_AGRI_ZS_UN','WB_ARD__AG_LND_TOTL_K2',
         'WB_ARD__SL_AGR_EMPL_ZS']
fields_3 = list(country_features.keys())
#for x in fields_3:
#    df = country_features[x]
#    if x in fields:
#        print('\033[94m'+x+': '+'\033[92m'+A_country.description[x]+' has '+str(sum(df.isna()))+' NA '+'\033[91m'+str(sum(df.isna())/len(df)))
#    else:
#        print('\033[94m'+x+': '+'\033[0m'+A_country.description[x]+' has '+str(sum(df.isna()))+' NA '+'\033[91m'+str(sum(df.isna())/len(df)))
nas = []
nas_per = []
des = []
for x in fields_3:
    df = country_features[x]
    nas.append(sum(df.isna()))
    nas_per.append(sum(df.isna())/len(df))
    des.append(A_country.description[x])

    ###Printing the ones with NA < 10%
DF = pd.DataFrame({'Indicator': fields_3, 'Description': des, 'Percentage of NA':nas_per})
DF[DF["Percentage of NA"]<0.1]

Indicator  \
116    WB_SE__WB_NY_GDP_MKTP_KD_ZG   
117    WB_SE__WB_NY_GDP_MKTP_PP_KD   
158           TI_CPI__TI_CPI_CPIRN   
159           TI_CPI__TI_CPI_CPIRO   
160           TI_CPI__TI_CPI_CPISN   
161           TI_CPI__TI_CPI_CPISO   
162         WB_ARD__AG_AGR_TRAC_NO   
164         WB_ARD__AG_CON_FERT_ZS   
165         WB_ARD__AG_LND_AGRI_K2   
166         WB_ARD__AG_LND_AGRI_ZS   
167         WB_ARD__AG_LND_ARBL_HA   
168      WB_ARD__AG_LND_ARBL_HA_PC   
169         WB_ARD__AG_LND_ARBL_ZS   
170         WB_ARD__AG_LND_CREL_HA   
171         WB_ARD__AG_LND_CROP_ZS   
174         WB_ARD__AG_LND_FRST_K2   
175         WB_ARD__AG_LND_FRST_ZS   
177         WB_ARD__AG_LND_PRCP_MM   
178         WB_ARD__AG_LND_TOTL_K2   
180         WB_ARD__AG_LND_TRAC_ZS   
181         WB_ARD__AG_PRD_CREL_MT   
182         WB_ARD__AG_PRD_CROP_XD   
183         WB_ARD__AG_PRD_FOOD_XD   
184         WB_ARD__AG_PRD_LVSK_XD   
185         WB_ARD__AG_SRF_TOTL_K2   
186         WB_ARD__AG_YLD_CREL_KG   
187      WB_ARD__EG_ELC_ACCS_RU_ZS   
188   WB_ARD__EN_ATM_METH_AG_KT_CE   
189      WB_ARD__EN_ATM_METH_AG_ZS   
190   WB_ARD__EN_ATM_NOXE_AG_KT_CE   
..                             ...   
327      WB_SPL__SL_UEM_TOTL_FE_ZS   
328   WB_SPL__SL_UEM_TOTL_MA_NE_ZS   
329      WB_SPL__SL_UEM_TOTL_MA_ZS   
330      WB_SPL__SL_UEM_TOTL_NE_ZS   
331         WB_SPL__SL_UEM_TOTL_ZS   
362    WB_SD__SE_ADT_1524_LT_FM_ZS   
363       WB_SD__SE_ENR_PRIM_FM_ZS   
364       WB_SD__SE_ENR_PRSC_FM_ZS   
365       WB_SD__SE_ENR_SECO_FM_ZS   
366       WB_SD__SE_ENR_TERT_FM_ZS   
367          WB_SD__SG_GEN_PARL_ZS   
368       WB_SD__SH_HIV_1524_FE_ZS   
369       WB_SD__SH_HIV_1524_MA_ZS   
370       WB_SD__SL_EMP_VULN_FE_ZS   
371       WB_SD__SL_EMP_VULN_MA_ZS   
381  WB_SD__SL_TLF_ACTI_1524_FE_ZS   
382  WB_SD__SL_TLF_ACTI_1524_MA_ZS   
383     WB_SD__SL_TLF_ACTI_1524_ZS   
384       WB_SD__SL_TLF_ACTI_FE_ZS   
385       WB_SD__SL_TLF_ACTI_MA_ZS   
386          WB_SD__SL_TLF_ACTI_ZS   
387       WB_SD__SL_TLF_CACT_FE_ZS   
388       WB_SD__SL_TLF_CACT_MA_ZS   
389       WB_SD__SL_UEM_TOTL_FE_ZS   
390       WB_SD__SL_UEM_TOTL_MA_ZS   
391             WB_SD__SM_POP_REFG   
392          WB_SD__SM_POP_REFG_OR   
393             WB_SD__SP_ADO_TFRT   
394       WB_SD__SP_DYN_LE00_FE_IN   
395       WB_SD__SP_DYN_LE00_MA_IN   

                                           Description  Percentage of NA  
116                              GDP growth (annual %)          0.036082  
117           GDP, PPP (constant 2011 international $)          0.061856  
158  Corruption Perception Index - Country Ranking ...          0.020619  
159  Corruption Perception Index - Country Ranking ...          0.024485  
160  Corruption Perception Index - Score (New Metho...          0.020619  
161  Corruption Perception Index - Score (Old Metho...          0.024485  
162                   Agricultural machinery, tractors          0.070876  
164  Fertilizer consumption (kilograms per hectare ...          0.092784  
165                         Agricultural land (sq. km)          0.014175  
166                 Agricultural land (% of land area)          0.021907  
167                             Arable land (hectares)          0.021907  
168                  Arable land (hectares per person)          0.021907  
169                       Arable land (% of land area)          0.021907  
170            Land under cereal production (hectares)          0.018041  
171                Permanent cropland (% of land area)          0.025773  
174                               Forest area (sq. km)          0.014175  
175                       Forest area (% of land area)          0.021907  
177       Average precipitation in depth (mm per year)          0.016753  
178                                 Land area (sq. km)          0.021907  
180  Agricultural machinery, tractors per 100 sq. k...          0.072165  
181                    Cereal production (metric tons)          0.018041  
182            Cro

In [12]:
fields = ['WB_ARD__AG_SRF_TOTL_K2', 'WB_SE__WB_NY_GDP_MKTP_KD_ZG', 'WB_ARD__SP_RUR_TOTL', 'WB_ARD__EG_ELC_ACCS_RU_ZS', 'WB_ARD__AG_LND_TOTL_RU_K2',
          'WB_ARD__AG_LND_FRST_K2', 'WB_ARD__AG_LND_ARBL_ZS', 'WB_ARD__AG_LND_AGRI_K2', 'WB_ARD__AG_CON_FERT_ZS','WB_ARD__AG_LND_PRCP_MM','WB_ARD__EN_ATM_METH_AG_KT_CE',
         'WB_ARD__EN_ATM_NOXE_AG_KT_CE','WB_ARD__NV_AGR_TOTL_ZS','WB_ARD__TM_VAL_AGRI_ZS_UN','WB_ARD__TX_VAL_AGRI_ZS_UN','WB_ARD__AG_LND_TOTL_K2',
         'WB_ARD__SL_AGR_EMPL_ZS']
imputed_ = []
for x in fields:
    df_temp = country_features[x]
    imputed_.append(fast_imputation(df_temp))

In [13]:
all_points_new = copy.deepcopy(all_points)

#### New dataset

In [14]:
for i in range(len(fields)):
    x = fields[i]
    all_points_new[x] = imputed_[i]

all_points_new.to_pickle('./data/all_points_new.pkl')

In [15]:
all_points_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22107 entries, 0 to 19999
Data columns (total 23 columns):
lat                             22107 non-null object
long                            22107 non-null float64
country                         22107 non-null object
year                            22107 non-null object
positive                        22107 non-null int64
random                          22107 non-null int64
WB_ARD__AG_SRF_TOTL_K2          1525 non-null float64
WB_SE__WB_NY_GDP_MKTP_KD_ZG     1525 non-null float64
WB_ARD__SP_RUR_TOTL             1525 non-null float64
WB_ARD__EG_ELC_ACCS_RU_ZS       1525 non-null float64
WB_ARD__AG_LND_TOTL_RU_K2       1525 non-null float64
WB_ARD__AG_LND_FRST_K2          1525 non-null float64
WB_ARD__AG_LND_ARBL_ZS          1525 non-null float64
WB_ARD__AG_LND_AGRI_K2          1525 non-null float64
WB_ARD__AG_CON_FERT_ZS          1525 non-null float64
WB_ARD__AG_LND_PRCP_MM          1525 non-null float64
WB_ARD__EN_ATM_METH_AG_KT_CE

#### The columns added are (in order):

In [16]:
for x in fields:
    print(A_country.description[x])

Surface area (sq. km)
GDP growth (annual %)
Rural population
Access to electricity, rural (% of rural population)
Rural land area (sq. km)
Forest area (sq. km)
Arable land (% of land area)
Agricultural land (sq. km)
Fertilizer consumption (kilograms per hectare of arable land)
Average precipitation in depth (mm per year)
Agricultural methane emissions (thousand metric tons of CO2 equivalent)
Agricultural nitrous oxide emissions (thousand metric tons of CO2 equivalent)
Agriculture, forestry, and fishing, value added (% of GDP)
Agricultural raw materials imports (% of merchandise imports)
Agricultural raw materials exports (% of merchandise exports)
Land area (sq. km)
Employment in agriculture (% of total employment) (modeled ILO estimate)


#### Also this paper https://academic.oup.com/bioinformatics/article/21/8/1509/249540, suggests that for highly correlated features the optimal feature size is
#### sqrt(N), where N is the sample size. For us this would be roughly 116 features.. 

In [18]:
gdp = country_features['WB_SE__WB_NY_GDP_MKTP_KD_ZG']

4.834865000000001